python/controller_selector.py

```diff
@@ -0,0 +1,268 @@
+                self.__switchFromUnknown(boat_speed)
+                return True
+
+        return False
+
+    def getControlMode(self):
+        """
+        Returns
+        -------
+        Returns a class reference to the current control mode used by the rudder.
+
+        """
+        return self.__controlMode
+
+    def getControlModeID(self):
+        """
+        Returns
+        -------
+        int
+            Returns an integer representing the control mode ID
+        """
+        return self.__controlModeID
+
+    def __switchFromUnknown(self, boat_speed):
+        """
+        Switches the controller mode from the UNKNOWN controller to another controller.
+
+        Switching Conditions
+        --------------------
+        JIBE_ONLY:
+            - The boat speed is less than or equal to a speed threshold.
+
+        TACKABLE:
+            - Default control mode. If the switch condition is not met for any other
+              controller, then assign the control mode to TACKABLE.
+
+        Arguments
+        ---------
+        float : boat_speed
+            The current boat speed in knots.
+
+        """
+        if(boat_speed <= sailbot_constants.SPEED_THRESHOLD_FOR_JIBING_KNOTS):
+            self.__assignMode(ControlModes.JIBE_ONLY.value)
+
+        else:
+            self.__assignMode(ControlModes.TACKABLE.value)
+
+        return
+
+    def __switchFromTacking(self, boat_speed, current_time, latest_switch_time, heading_error):
+        """
+        The controller switches from TACKABLE mode to UNKNOWN mode if it meets its
+        exit condition. The boat must be in TACKABLE mode when this method is invoked
+        or there may be unexpected results.
+
+        Exit Condition
+        --------------
+        - The boat speed is less than or equal to the speed threshold, OR
+        - The boat has been in TACKABLE mode for longer than the specified timeout, OR
+        - The boat has reached the specified setpoint
+
+        Arguments
+        ---------
+        float : boat_speed
+            The current boat speed in knots.
+
+        int : current_time
+            The current time (unix timestamp)
+
+        int : latest_switch_time
+            The time of the latest switch of the controller mode
+
+        float : heading_error
+            The current heading error in radians
+
+        """
+        tooSlow = boat_speed <= sailbot_constants.SPEED_THRESHOLD_FOR_JIBING_KNOTS
+        timeout = current_time - latest_switch_time >= sailbot_constants.MAX_TIME_FOR_TACKING
+        reached_heading = abs(heading_error) <= sailbot_constants.MIN_HEADING_ERROR_FOR_SWITCH
+
+        # TODO: Panic mode. If we are not at the desired heading by the end of the timeout then
+        #       we need to do something else.
+        if(tooSlow or timeout or reached_heading):
+            self.__assignMode(ControlModes.UNKNOWN.value)
+        return
+
+    def __switchFromJibing(self, current_time, latest_switch_time, heading_error):
```

Comment:
just a general comment, if you want we don't have to be as thorough for documenting private methods as they are more prone to changes. I personally just make the method name more descriptive/longer